rskj-core/src/main/java/co/rsk/core/bc/ParallelizeTransactionHandler.java

```diff
@@ -0,0 +1,249 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2017 RSK Labs Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package co.rsk.core.bc;
+
+import co.rsk.core.RskAddress;
+import org.ethereum.core.Transaction;
+import org.ethereum.db.ByteArrayWrapper;
+import org.ethereum.vm.GasCost;
+
+import java.util.*;
+
+public class ParallelizeTransactionHandler {
+    private final TransactionBucket sequentialBucket;
+    private final HashMap<ByteArrayWrapper, Set<Short>> bucketByWrittenKey;
+    private final HashMap<ByteArrayWrapper, Set<Short>> bucketByReadKey;
+    private final Map<RskAddress, Set<Short>> bucketBySender;
+    private final List<TransactionBucket> parallelBuckets;
+
+    public ParallelizeTransactionHandler(short buckets, long blockGasLimit) {
+        long gasLimit = blockGasLimit/2;
+        this.bucketBySender = new HashMap<>();
+        this.bucketByWrittenKey = new HashMap<>();
+        this.bucketByReadKey = new HashMap<>();
+        this.parallelBuckets = new ArrayList<>();
+        for (short i = 0; i < buckets; i++){
+            this.parallelBuckets.add(new TransactionBucket(i, gasLimit));
+        }
+        this.sequentialBucket = new TransactionBucket(buckets, gasLimit);
+    }
+
+    public boolean sequentialBucketHasGasAvailable(Transaction tx) {
+        long txGasLimit = GasCost.toGas(tx.getGasLimit());
+        return this.sequentialBucket.hasGasAvailable(txGasLimit);
+    }
+
+    public Optional<Short> addTransaction(Transaction tx, Set<ByteArrayWrapper> newReadKeys, Set<ByteArrayWrapper> newWrittenKeys, long gasUsedByTx) {
```

Comment:
should this throw an exception if there is not enough gas in the sequential bucket?

Reply:
Line 49 is only a query for the available gas. Then, when the handler tries to add a transaction into the sequential bucket, if the sequential has no gas, it returns an `Optional.empty()`.